File: app/src/main/java/example/com/englishnote/VocaListActivity.java
package example.com.englishnote;

import android.content.Intent;
import android.os.Bundle;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.room.Room;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import example.com.englishnote.adapter.VocaAdapter;
import example.com.englishnote.common.IntentExtra;
import example.com.englishnote.database.AppDatabase;
import example.com.englishnote.model.Vocabulary;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class VocaListActivity extends AppCompatActivity {
    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.list_voca)
    RecyclerView vocaList;

    private AppDatabase mDb;
    private Observable<List<Vocabulary>> mVocaListObservable;
    private Disposable mVocaListDisposable;

    private VocaAdapter mAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_voca_list);
        ButterKnife.bind(this);

        ActionBarManager.initBackArrowActionbar(this, toolbar, getString(R.string.title_voca_list));
        mDb = Room.databaseBuilder(getApplicationContext(), AppDatabase.class,
                AppDatabase.DATABASE_NAME).build();
    }

    @Override
    protected void onStart() {
        super.onStart();
        observeVocaList();
    }

    private void setRecyclerView(@NonNull List<Vocabulary> data) {
        vocaList.setLayoutManager(new LinearLayoutManager(this));
        mAdapter = new VocaAdapter(data);
        vocaList.setAdapter(mAdapter);
    }

    protected void observeVocaList() {
        mVocaListDisposable = mDb.vocaDao().selectAll().observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(vocaList -> setRecyclerView(vocaList));
    }

    private void goToMainActivity() {
        finish();
    }

    @OnClick(R.id.button_add)
    public void onAddBtnClicked() {
        startActivity(new Intent(this, EditVocaActivity.class)
                .putExtra(IntentExtra.VOCA_ID, IntentExtra.VOCA_NULL));
    }

    @Override
    protected void onStop() {
        super.onStop();
        mVocaListDisposable.dispose();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        if (id == R.id.home || id == android.R.id.home) {
            goToMainActivity();
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {
        goToMainActivity();
    }
}

File: app/src/main/java/com/example/wordbook/vocalist/VocaListActivity.kt
package com.example.wordbook.vocalist

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.wordbook.R
import com.example.wordbook.database.Word
import com.example.wordbook.databinding.ActivityVocaListBinding
import com.example.wordbook.edit.EditVocaActivity
import com.example.wordbook.register.RegisterVocaActivity
import com.example.wordbook.util.Constants

class VocaListActivity : AppCompatActivity() {
    private lateinit var mViewModel : VocaListViewModel
    private lateinit var mBinding: ActivityVocaListBinding
    private lateinit var mAdapter: VocaListAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        mViewModel = ViewModelProvider(this).get(VocaListViewModel::class.java)
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_voca_list)

        mBinding.viewModel = mViewModel
        mBinding.list.layoutManager = LinearLayoutManager(this)
        mAdapter = VocaListAdapter(::moveToEditVoca)
        mBinding.list.adapter = mAdapter

        mViewModel.moveToRegisterVoca.observe(this) {
            if (it) {
                moveToRegisterVoca()
            }
        }

        mViewModel.vocas.observe(this) {
            mAdapter.submitList(it)
        }
    }

    private fun moveToEditVoca(word: Word) {
        val intent = Intent(this, EditVocaActivity::class.java)
        intent.putExtra(Constants.EXTRA_VOCA_ID, word?.id ?: -1)
        startActivity(intent)
    }

    private fun moveToRegisterVoca() {
        startActivity(Intent(this, RegisterVocaActivity::class.java))
    }
}